platform/ios/src/MGLMapView.mm

```diff
@@ -3850,6 +3952,75 @@ - (void)updateUserLocationAnnotationView
     [self updateUserLocationAnnotationViewAnimatedWithDuration:0];
 }
 
+- (void)updateAnnotationViews
+{
+    BOOL delegateImplementsViewForAnnotation = [self.delegate respondsToSelector:@selector(mapView:viewForAnnotation:)];
+    
+    if (!delegateImplementsViewForAnnotation)
+    {
+        return;
+    }
+    
+    // Update all visible annotation views
+    std::set<MGLAnnotationTag> visibleTags;
+    std::vector<MGLAnnotationTag> annotationTags = [self annotationTagsInRect:self.bounds];
+    
+    for(auto const& annotationTag: annotationTags)
+    {
+        auto &annotationContext = _annotationContextsByAnnotationTag[annotationTag];
+        id<MGLAnnotation> annotation = annotationContext.annotation;
+        
+        CGPoint annotationPoint = [self convertCoordinate:annotation.coordinate toPointToView:self];
+       
+        // If there is no annotation view at this point, it means the context's view was reused by some
+        // other context so we need to reuse or make a new view.
+        if (!annotationContext.annotationView)
+        {
+            MGLAnnotationView *annotationView = [self annotationViewForAnnotation:annotation];
+            
+            if (annotationView)
+            {
+                // If the annotation view has no superview it means it was never used before so add it
+                if (!annotationView.superview)
+                {
+                    [self.glView addSubview:annotationView];
+                }
+                
+                annotationContext.annotationView = annotationView;
+            }
+        }
+       
+        annotationContext.annotationView.center = annotationPoint;
+        annotationContext.annotationView.hidden = NO;
+        
+        visibleTags.insert(annotationTag);
+    }
+    
+    // Hide and add offscreen annotation views to reuse queue
+    for (auto &pair : _annotationContextsByAnnotationTag)
+    {
+        MGLAnnotationTag annotationTag = pair.first;
+        MGLAnnotationContext &annotationContext = pair.second;
+        MGLAnnotationView *annotationView = annotationContext.annotationView;
+        const bool tagIsNotVisible = visibleTags.find(annotationTag) == visibleTags.end();
+        
+        if (annotationView && tagIsNotVisible)
+        {
+            annotationView.hidden = YES;
+            
+            if (annotationContext.viewReuseIdentifier)
+            {
+                NSMutableArray *annotationViewReuseQueue = [self annotationViewReuseQueueForIdentifier:annotationContext.viewReuseIdentifier];
+                if (![annotationViewReuseQueue containsObject:annotationView])
```

Comment:
It's a pointer comparison. This guard just protects against adding a reference to any exact same view object that was previously added to the queue. I used an array for the queue since that API feels better (as a queue) in most places -- except here where a set might make things clearer. In any case, the intention is to capture any view object that goes offscreen in the appropriate queue so that it can be reused